android/src/main/java/com/tailscale/ipn/ui/notifier/Notifier.kt

```diff
@@ -47,95 +41,99 @@ class Notifier(private val scope: CoroutineScope) {
+    val version: StateFlow<String?> by lazy {
+        flow(null) { it.Version }
+    }
+
+    private fun <T> flow(initial: T, extract: (Notify) -> T?): StateFlow<T> {
+        val flow = MutableStateFlow(initial)
+        watchIPNBus(NotifyWatchOpt.netmap.value or NotifyWatchOpt.prefs.value or NotifyWatchOpt.initialState.value) { notify ->
+            extract(notify)?.let { flow.value = it }
         }
+        return flow
     }
 
-    // Starts an IPN Bus watcher.  **This is blocking** and will not return until
-    // the watcher is stopped and must be executed in a suitable coroutine scope such
-    // as Dispatchers.IO
+    // Starts an IPN Bus watcher. This is non-blocking.
     private external fun startIPNBusWatcher(sessionId: String, mask: Int)
 
-    // Stops an IPN Bus watcher
+    // Stops an IPN Bus watcher. This is non-blocking.
     private external fun stopIPNBusWatcher(sessionId: String)
 
-    private var watchers = HashMap<String, Watcher>()
-
     // Callback from jni when a new notification is received
-    fun onNotify(notification: String, sessionId: String) {
-        val notify = decoder.decodeFromString<Notify>(notification)
+    @OptIn(ExperimentalSerializationApi::class)
+    fun onNotify(notification: ByteArray, sessionId: String) {
+        val notify = decoder.decodeFromStream<Notify>(notification.inputStream())
         val watcher = watchers[sessionId]
-        watcher?.let { watcher.callback(notify) } ?: {
+        watcher?.let { watcher(notify) } ?: {
             Log.e(
                 "Notifier",
                 "Received notification for unknown session: ${sessionId}"
             )
         }
     }
 
+    fun close() {
+        watchers.keys.forEach { stopIPNBusWatcher(it) }
+        watchers.clear()
+    }
+
     // Watch the IPN bus for notifications
     // Notifications will be passed to the caller via the callback until
     // the caller calls unwatchIPNBus with the sessionId returned from this call.
-    private fun watchIPNBus(mask: Int, callback: NotifierCallback): String {
-        val sessionId = generateSessionId()
-        val watcher = Watcher(sessionId, mask, callback)
-        watchers[sessionId] = watcher
+    private fun watchIPNBus(mask: Int, callback: (Notify) -> Unit): String {
+        val sessionId = UUID.randomUUID().toString()
+        watchers[sessionId] = callback
         scope.launch(Dispatchers.IO) {
             // Wait for the notifier to be ready
             isReady.await()
-            Log.d("Notifier", "Starting IPN Bus watcher for sessionid: ${sessionId}")
+            Log.d(
+                "Notifier",
+                "Starting IPN Bus watcher for sessionid: ${sessionId} in job ${scope.coroutineContext.job}"
+            )
             startIPNBusWatcher(sessionId, mask)
-            watchers.remove(sessionId)
-            Log.d("Notifier", "IPN Bus watcher for sessionid:${sessionId} has halted")
+            scope.coroutineContext.job.invokeOnCompletion {
```

Comment:
This ties us to the scope's lifecycle, so that if the scope's job is canceled/finishes, we'll clean up watcher.